aptos-move/framework/src/natives/aggregator_natives/aggregator_snapshot.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::natives::aggregator_natives::{
+    helpers_v2::{aggregator_info, aggregator_snapshot_u128_info, aggregator_snapshot_u64_info},
+    NativeAggregatorContext,
+};
+use aptos_aggregator::aggregator_extension::AggregatorSnapshotID;
+use aptos_gas_schedule::gas_params::natives::aptos_framework::{
+    AGGREGATOR_V2_READ_SNAPSHOT_BASE, AGGREGATOR_V2_SNAPSHOT_BASE,
+    AGGREGATOR_V2_SNAPSHOT_WITH_U64_LIMIT_BASE,
+};
+use aptos_native_interface::{
+    safely_pop_arg, RawSafeNative, SafeNativeBuilder, SafeNativeContext, SafeNativeResult,
+};
+use move_vm_runtime::native_functions::NativeFunction;
+use move_vm_types::{
+    loaded_data::runtime_types::Type,
+    values::{Struct, StructRef, Value},
+};
+use smallvec::{smallvec, SmallVec};
+use std::collections::VecDeque;
+
+/***************************************************************************************************
+ * native fun snapshot(aggregator: &Aggregator): AggregatorSnapshot<u128>;
+ **************************************************************************************************/
+
+fn native_snapshot(
+    context: &mut SafeNativeContext,
+    _ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(args.len(), 1);
+
+    context.charge(AGGREGATOR_V2_SNAPSHOT_BASE)?;
+    let (aggregator_id, _limit) = aggregator_info(&safely_pop_arg!(args, StructRef))?;
+
+    // Get aggregator.
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+    let AggregatorSnapshotID { id } = aggregator_data.snapshot(&aggregator_id);
+
+    Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u128(
+        id as u128
+    ),]))])
+}
+
+/***************************************************************************************************
+ * native fun snapshot_with_u64_limit(aggregator: &Aggregator): AggregatorSnapshot<u64>;
+ **************************************************************************************************/
+
+fn native_snapshot_with_u64_limit(
+    context: &mut SafeNativeContext,
+    _ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(args.len(), 1);
+
+    context.charge(AGGREGATOR_V2_SNAPSHOT_WITH_U64_LIMIT_BASE)?;
+    let (aggregator_id, _limit) = aggregator_info(&safely_pop_arg!(args, StructRef))?;
```

Comment:
you should check that limit is <= u64::MAX, and if it is not , abort

Reply:
So, we can't crate a snapshot_u64 if the limit is above u64::MAX?

Reply:
Indeed, if I have a limit of u128::max, but my aggregator stores 10, why cannot I create a snapshot_u64...
Unless we want to be restrictive here, as allowing the above means we only check overflow on reading / materializing the snapshot.

Reply:
My suggestion is that we don't want code to sometimes be able to cast to u64, sometimes not,  and then need to handle that too. 
And second reason is that since limit is fixed, casting success or abort is never wrong in speculation, and so casting never triggers re-execution

Basically I think this should work as if Aggregator was a templated on type, and there is no casting - u64 aggregator creates u64 snapshot, u128 aggregator creates u128 snapshot. 

I think having Aggregator be a template is more complicated,  but if it isn't,  maybe we should do that?

Reply:
Given that aggregator can never be replaced in backward compatible way having only u128 version of aggregator seems simpler


But what I meant to say,  above suggestion makes it that casting to u64 should never fail,  if it wasn't a coding error - i.e. if you need to cast to u64 snapshot,  you should assert limit is within u64 bounds before you create an aggregator, and you should be good to go.